### PR TITLE
add IAMAuthPolicy watchers for Gateway and Routes

### PR DESCRIPTION
**Pull request by mikhail-aws**
Note:
Add IAMAuthPolicy watchers for Gateway and Routes.

Tested manually with creation and deletion of Gateway, Route, and IAMAuthPolicy in different order.
If GW or Route does not exists IAMAuthPolicy status will be TargetNotFound. Once GW or Route is created, policy will eventually be there, after few rounds of reconciliation.

Listening GW events is easy, and does not trigger redundant reconciles for Auth policy. But Route updates are massive, we need to revisit Route logic and how we update K8s Status or Spec, otherwise our watchers generate 4-5 times extra load for each route update.

I tried to use GenerationChangedPredicate for Route events, but it didnt work for me. Route mostly update K8s Status, which is filtered out by GenerationChangedPredicate. I dont know why/how it works for AccessLog, or does it really work?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: controllers/iamauthpolicy_controller.go [hunk 1]
```diff
@@ -249,13 +275,50 @@ func (c IAMAuthPolicyController) updatePolicyCondition(ctx context.Context, k8sP
...
-		Type:               string(gwv1alpha2.PolicyConditionAccepted),
-		Status:             status,
-		ObservedGeneration: k8sPolicy.Generation,
-		LastTransitionTime: metav1.Now(),
-		Reason:             string(reason),
+		Type:   string(gwv1alpha2.PolicyConditionAccepted),
+		Status: status,
+		Reason: string(reason),
 	}
 	meta.SetStatusCondition(&k8sPolicy.Status.Conditions, cnd)
```
Comment (zijun726911) [inline controllers/iamauthpolicy_controller.go:282]: Off topic: How you figured out to use `meta.SetStatusCondition()`? Seems elsewhere in our repo should also it to update status...

https://github.com/aws/aws-application-networking-k8s/blob/a2d5b973b3516f11784d1a764830e8de2eadf4a7/controllers/accesslogpolicy_controller.go#L302

https://github.com/aws/aws-application-networking-k8s/blob/cd77f5d506cc9ecc354dc46460f0364188113129/pkg/utils/condition.go#L7

File: controllers/iamauthpolicy_controller.go [hunk 2]
```diff
@@ -249,13 +275,50 @@ func (c IAMAuthPolicyController) updatePolicyCondition(ctx context.Context, k8sP
...
+
+func (c *IAMAuthPolicyController) handleTargetRefNotFound(ctx context.Context, k8sPolicy *anv1alpha1.IAMAuthPolicy, err error) error {
+	if services.IsNotFoundError(err) {
+		err = c.updatePolicyCondition(ctx, k8sPolicy, gwv1alpha2.PolicyReasonTargetNotFound)
+	}
+	return err
+}
+
+func (c *IAMAuthPolicyController) updateLatticeAnnotaion(ctx context.Context, k8sPolicy *anv1alpha1.IAMAuthPolicy, resId string) error {
+	k8sPolicy.Annotations["application-networking.k8s.aws/resourceId"] = resId
```
Comment (zijun726911) [inline controllers/iamauthpolicy_controller.go:321]: Define a constant for Annotations key?

https://github.com/aws/aws-application-networking-k8s/blob/a2d5b973b3516f11784d1a764830e8de2eadf4a7/pkg/apis/applicationnetworking/v1alpha1/accesslogpolicy_types.go#L13

File: controllers/iamauthpolicy_controller.go [hunk 3]
```diff
@@ -249,13 +275,50 @@ func (c IAMAuthPolicyController) updatePolicyCondition(ctx context.Context, k8sP
...
+	return func(obj client.Object) []ctrl.Request {
+		requests := []ctrl.Request{}
+		policies := &anv1alpha1.IAMAuthPolicyList{}
+		err := c.List(context.Background(), policies, &client.ListOptions{Namespace: obj.GetNamespace()})
+		if err != nil {
+			log.Error(err)
+			return requests
+		}
+		for _, policy := range policies.Items {
+			if obj.GetName() == string(policy.Spec.TargetRef.Name) {
```
Comment (zijun726911) [inline controllers/iamauthpolicy_controller.go:297]: Do you need to check `objKind == policy.Spec.TargetRef.Kind` ?

https://github.com/aws/aws-application-networking-k8s/blob/a2d5b973b3516f11784d1a764830e8de2eadf4a7/controllers/accesslogpolicy_controller.go#L346-L349

File: controllers/iamauthpolicy_controller.go [hunk 4]
```diff
@@ -47,17 +60,9 @@ func (c *IAMAuthPolicyController) Reconcile(ctx context.Context, req ctrl.Reques
 	}
 	c.log.Infow("reconcile", "req", req, "targetRef", k8sPolicy.Spec.TargetRef)
```
Comment (zijun726911) [inline controllers/iamauthpolicy_controller.go:62]: Need some guard clauses that similar to these?

https://github.com/aws/aws-application-networking-k8s/blob/a2d5b973b3516f11784d1a764830e8de2eadf4a7/controllers/accesslogpolicy_controller.go#L129-L145

File: controllers/iamauthpolicy_controller.go [hunk 5]
```diff
@@ -139,21 +139,32 @@ func (c *IAMAuthPolicyController) findSnId(ctx context.Context, k8sPolicy *anv1a
 	return *snInfo.SvcNetwork.Id, nil
 }
 
-func (c *IAMAuthPolicyController) upsertGatewayPolicy(ctx context.Context, k8sPolicy *anv1alpha1.IAMAuthPolicy) (string, error) {
+func (c *IAMAuthPolicyController) upsertGatewayPolicy(ctx context.Context, k8sPolicy *anv1alpha1.IAMAuthPolicy) error {
+	err := c.handleConflicts(ctx, k8sPolicy)
+	if err != nil {
+		return err
+	}
 	snId, err := c.findSnId(ctx, k8sPolicy)
```
Comment (zijun726911) [inline controllers/iamauthpolicy_controller.go:147]: Just check LatticeSN/Serve resource to determine whether update Policy status to `gwv1alpha2.PolicyReasonTargetNotFound` looks a smart way, I think it works well?

File: controllers/iamauthpolicy_controller.go (hunk 1 above)
Comment (mikhail-aws) [reply controllers/iamauthpolicy_controller.go:282]: I searched for the utility methods and found it. I dont think we need our own utility functions.

File: controllers/iamauthpolicy_controller.go (hunk 2 above)
Comment (mikhail-aws) [reply controllers/iamauthpolicy_controller.go:321]: it's not used anywhere else

File: controllers/iamauthpolicy_controller.go (hunk 3 above)
Comment (mikhail-aws) [reply controllers/iamauthpolicy_controller.go:297]: when we create create a "watcher" we already declare which kind of resource to watch, this condition is redundant

File: controllers/iamauthpolicy_controller.go (hunk 4 above)
Comment (mikhail-aws) [reply controllers/iamauthpolicy_controller.go:62]: I can add group check, kind check is already there. https://github.com/aws/aws-application-networking-k8s/blob/main/controllers/iamauthpolicy_controller.go#L61-L77

File: controllers/iamauthpolicy_controller.go (hunk 5 above)
Comment (mikhail-aws) [reply controllers/iamauthpolicy_controller.go:147]: didnt get the question

File: controllers/iamauthpolicy_controller.go (hunk 1 above)
Comment (zijun726911) [reply controllers/iamauthpolicy_controller.go:282]: Great, I created a issue for it https://github.com/aws/aws-application-networking-k8s/issues/459

File: controllers/iamauthpolicy_controller.go (hunk 5 above)
Comment (zijun726911) [reply controllers/iamauthpolicy_controller.go:147]: The [accessLogPolicy](https://github.com/aws/aws-application-networking-k8s/blob/a2d5b973b3516f11784d1a764830e8de2eadf4a7/controllers/accesslogpolicy_controller.go#L214), check ref k8s resource exist or not to update alp with PolicyReasonTargetNotFound, (and check lattice resource later again in buildAndDeployModel and [do nothing if latticeResource not found?](https://github.com/aws/aws-application-networking-k8s/blob/de8994c8ffbab778f64ab9fe26f3e82f0bdfa6da/pkg/deploy/lattice/access_log_subscription_manager.go#L46), I didn't quite get this part ALP logic)

Anyway, I think this PR this part is good

**Review by zijun726911 — APPROVED** [review]
LGTM, for minor issues, we could fix them afterward in seperate PRs

File: controllers/iamauthpolicy_controller.go (hunk 5 above)
Comment (mikhail-aws) [reply controllers/iamauthpolicy_controller.go:147]: good point, probably PolicyReasonTargetNotFound should only check for k8s reference not lattice

**Pull request merged**